pkg/controller/job/job_controller.go

```diff
@@ -232,9 +232,14 @@ func (jm *Controller) Run(ctx context.Context, workers int) {
 
 	for i := 0; i < workers; i++ {
 		go wait.UntilWithContext(ctx, jm.worker, time.Second)
+		go wait.UntilWithContext(ctx, jm.orphanWorker, time.Second)
 	}
 
-	go wait.UntilWithContext(ctx, jm.orphanWorker, time.Second)
+	// Even if the jobs pods are not tracked by the main reconciliation loop.
+	// Make sure the Pod would be cleaned up by the orphan pod worker, when the Pod finishes.
+	if workers == 0 {
+		go wait.UntilWithContext(ctx, jm.orphanWorker, time.Second)
+	}
```

Comment:
As you noted in our slack conversation workers in production will rarely be 0, it's only in our unit tests, where we force it for specific reason. We even have a check in https://github.com/kubernetes/kubernetes/blob/0a406ee1b4b65574c02f8aec6ff5a69d07b68be6/cmd/kube-controller-manager/app/options/jobcontroller.go#L53-L63 which will ensure it's a positive number. So it's wrong to do it here. You need to update the test, like I mentioned before.